crates/matrix-sdk-base/src/client.rs

```diff
@@ -709,58 +720,108 @@ impl BaseClient {
 
             changes.add_room(room_info);
         }
+        Ok(())
+    }
 
-        for (room_id, new_info) in rooms.leave {
-            let room = self.store.get_or_create_room(&room_id, RoomType::Left).await;
-            let mut room_info = room.clone_info();
-            room_info.mark_as_left();
+    /// Receive a response from a sync call.
+    ///
+    /// This will additionally to `process_sync_response` also check if we
+    /// already have process the sync token before and if so, we skip it.
+    ///
+    /// # Arguments
+    ///
+    /// * `response` - The response that we received after a successful sync.
+    /// * `transaction` - If true we got an appservice transaction instead of a
+    ///   sync.
+    pub async fn receive_sync_response(
+        &self,
+        response: api::sync::sync_events::v3::Response,
+        transaction: bool,
+    ) -> Result<SyncResponse> {
+        #[allow(unused_variables)]
+        let api::sync::sync_events::v3::Response {
+            ref next_batch,
+            rooms,
+            presence,
+            account_data,
+            to_device,
+            device_lists,
+            device_one_time_keys_count,
+            device_unused_fallback_key_types,
+            ..
+        } = response;
+        if transaction {
+            // The server might send multiple times the same transaction id, in
+            // that case we already received this transaction and there's nothing to
+            // do.
+            //
+            // Spec: https://spec.matrix.org/v1.3/application-service-api/#pushing-events
+            if self.store.transaction_id.read().await.as_ref() == Some(next_batch) {
```

Comment:
I think his is the core issue I have with this PR: It's about not treating the transaction ID as equivalent to `next_batch`, but apparently it still comes in as exactly that here. Adding a `transaction` parameter to `receive_sync_response` seems like a bandaid, not a fix. Instead, I think appservices shouldn't be dealing with `sync_events::v3::Response` in the first place.

cc @poljar

Reply:
Hm yeah thats true. But then we get to the point where we need to basically rewrite major chunks of the sdk I think as the sync types are used in a lot more places.

I agree with this fully, however.

Reply:
How about just duplicating the relevant code that deals with sync responses and updating the duplicates to work on the appservice type? I think that would be a better starting point than this, from where things can then be de-duplicated.